### ISSUE TITLE
Training and validation, but what do we use to test?

### ISSUE THREAD

**Issue by odeniz** [vlfeat/matconvnet#161]
Hi All,

I use cnn_train for training and validation using an imdb structure with data. However, the data also contains a test part (the dataset contains train, val and test parts), which I understand is not being used by cnn_train. 

How do we get test error rates?

Thanks in advance

**Comment by odeniz** [comment]
In the MNIST and CIFAR examples cnn_train is used to perform validation on the test set (i.e.imdb.images.set == 3). Does this mean that some validation process is being performed on the test set (a possible bug)? or is it that validation only gets errors (i.e.with no effect on the resulting net) and therefore we can use this approach to get test error?

**Comment by bazilas** [comment]
Hi Oscar,

It's not a bug. These examples are demonstrating how the library works. If you would like to "properly" train the network, you should exclude the testing data and use a part of the training data for validation or the given validation data.

So, if you would like to plot the testing error (i.e. tune your network on the testing data), you would have to modify cnn_train(.) in order to run the network one more for the testing data and then modify the statistics / plots.


Vasilis

**Comment by odeniz** [comment]
Hi Vasilis,

I don't want to tune the network on testing data, only measure test error. It may not be a bug, but the examples lead to confusion when they use testing data for validation.
On the other hand, I've seen that inside cnn_train validation does nothing but obtaining error (no parameter seems to be tuned with the validation set). So what I suspect is that we can use this approach to get test error (that's what I want), i.e. by running cnn_train(....  'val', find(imdb.images.set==3)) ;

What do you think?

**Comment by bazilas** [comment]
Hi Oscar,

You could write another script (similar to cnn_train) only for the testing data. The important function is vl_simplenn(.) and some routines for preparing the data and network parameters. You could call the testing script to obtain your error and plot it. In some cases, you might want to export more than the error (e.g. predicted values). So it would be more convenient to isolate the testing part from the rest.

The fastest way as you wrote above is to use at the testing data as validation and get your error.

V

**Comment by anselmoferreira** [comment]
Hi people,

first of all thank you @bazilas and @odeniz for the discussion, this is exactly what I was looking for. I have the same problem reported by @odeniz, so, just to confirm:

1- can I use as 'validation' my testing data? 
2- no network parameters are changed using the 'validation', it is just a test in the network prepared in the given epoch using the 'training' set, am I right?

Again, thank you so much for the discussion and also for your attention,

**Comment by faraz107** [comment]
Hi @anselmoferreira 

I hope you were able to run the 'test data' as 'validation' for obtaining the test error rates in cnn_train

I am also looking to obtain test error rates and just trying to use the test data as validation.

Thanks,
